### PR TITLE
[Structural] Eigensolver missing nonlin function calls

### PR DESCRIPTION
**Pull request by philbucher**
the `EigenSolverStrategy` is missing the calls to `InitializeNonLinIteration` and `FinalizeNonLinIteration`
I am pretty convinced they should be there
They are also used in the Linear-Strategy

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/eigensolver_dynamic_scheme.hpp [hunk 1]
```diff
@@ -90,6 +90,50 @@ class EigensolverDynamicScheme : public Scheme<TSparseSpace,TDenseSpace>
     ///@name Operations
     ///@{
 
+    void InitializeNonLinIteration(
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/eigensolver_dynamic_scheme.hpp]: this can be removed after #5121

**Comment by armingeiser** [comment]
Why is this necessary? i mean, there are no non linear iterations happening?

**Comment by philbucher** [comment]
because the element might still do sth

Check the [linear strategy](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/solving_strategies/strategies/residualbased_linear_strategy.h#L598-L628), there it is done in the same way

**Comment by armingeiser** [comment]
> because the element might still do sth
> 
> Check the [linear strategy](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/solving_strategies/strategies/residualbased_linear_strategy.h#L598-L628), there it is done in the same way

Ok - i see why it might be necessary. No offence, but IMO this is either a hint that the design of the elements or the naming of this function is flawed...

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp [hunk 2]
```diff
@@ -402,6 +402,9 @@ class EigensolverStrategy
 
         rModelPart.GetProcessInfo()[BUILD_LEVEL] = 1;
         TSparseSpace::SetToZero(rMassMatrix);
+
+        pScheme->InitializeNonLinIteration(rModelPart, rMassMatrix, b, b); // arguments except ModelPart are unused in Scheme.
```
Comment (armingeiser) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: Even it is not used, it seems more natural to me to pass the `rStiffnessMatrix`. Up to you

**Review by armingeiser — DISMISSED** [review]

**Comment by philbucher** [comment]
> Ok - i see why it might be necessary. No offence, but IMO this is either a hint that the design of the elements or the naming of this function is flawed...

I agree

any comment @RiccardoRossi do you know why it is done like this?

**Comment by philbucher** [comment]
@emiroglu can you please check if this solves your problem?

**Comment by qaumann** [comment]
@philbucher 
> because the element might still do sth
> 
> Check the [linear strategy](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/solving_strategies/strategies/residualbased_linear_strategy.h#L598-L628), there it is done in the same way

Does this mean we always have to initialize a nonlinear iteration, even if the analysis we want to perform is linear? This is really confusing and I would expect rather a mistake if I saw some function regarding nonlinearities in a linear analysis strategy.

Further, are you implying that there are elements which cannot be used in a purely linear way at all?

**Comment by emiroglu** [comment]
> @emiroglu can you please check if this solves your problem?

Partially yes. So the flow in my case is as below:

```
EigenValueAnalysis()
for i_eigenvector in eigenvectors:
     UpdateNodalDOFs(i_eigenvector)
     EigenValueAnalysis()
```

So since the `FinalizeNonLinIteration()` is not called before UpdateNodalDOFs(i_eigenvector), the change in the system matrices is only effective starting from the second iteration of the loop. I think one needs to have a similar function to `FinalizeNonLinIteration()` such that the element internal quantities are updated before the analysis. 

How is it even done for the restarted analyses?

**Comment by philbucher** [comment]
> This is really confusing and I would expect rather a mistake if I saw some function regarding nonlinearities in a linear analysis strategy.

I agree

This is related to  #5121 
If you check closely for some schemes the `InitializeNonLinIteration` is done inside the `CalculateLocalSystem` functions :D

**Comment by philbucher** [comment]
@RiccardoRossi what do you think abt this?

**Comment by RiccardoRossi** [comment]
the scenario i am thinking of is when one implements a nonlinear strategy by calling multiple times a linear SolveSoltutionStep.

imagine of doing

     linear_strategy.InitializeSolutionStep(...)
     while(not converged)
           linear_strategy.SolveSolutionStep(...)
           ...
     linear_strategy.FinalizeSolutionStep(...)

a loop that is present e.g. in a simple staggered coupling scheme.

the only function call that we have here is the Initialize/Finalize nonlinear iteration

you might also want in some corner cases to use a nonlinear element within a linear strategy. And that element might implement that function and expect that it is called. Think for example of doing an eigenvalue analysis of a nonlinear element to compute the buckling load.

**Comment by philbucher** [comment]
@armingeiser @qaumann should we merge this? what do you think?

**Review by armingeiser — DISMISSED** [review]
Can be merged from my side

**Comment by philbucher** [comment]
I would merge this now since we are moving things towards the Strategies (see the new Explicit Strategy in the Core where those fcts are also called in the SolveSolutionStep)

**Comment by philbucher** [comment]
ok now is much cleaner (and depending on #8242)

**Comment by philbucher** [comment]
can someone reapprove please?

**Review by KlausBSautter — APPROVED** [review]

**Pull request merged**